torch/package/module_environment.py

```diff
@@ -0,0 +1,219 @@
+
+        Args:
+            obj: An object to get the the module-environment-relative name for.
+            name: If set, use this name instead of looking up __name__ or __qualname__ on `obj`.
+                This is only here to match how Pickler handles __reduce__ functions that return a string,
+                don't use otherwise.
+            check_obj: If true, check that retrieving the returned name produces the same object as `obj`.
+                This should only be false if you plan to call `check_obj()` somewhere else.
+
+        Returns:
+            A tuple (parent_module_name, attr_name) that can be used to retrieve `obj` from this environment.
+            Use it like:
+                mod = module_env.import_module(parent_module_name)
+                obj = getattr(mod, attr_name)
+
+        Raises:
+            ModuleEnvError: `check_obj()` has failed.
+        """
+        if name is None:
+            name = getattr(obj, "__qualname__", None)
+        if name is None:
+            name = obj.__name__
+
+        orig_module_name = self._whichmodule(obj, name)
+        # Demangle the module name before importing. If this obj came out of a
+        # PackageImporter, `__module__` will be mangled. See mangling.md for
+        # details.
+        module_name = demangle(orig_module_name)
+
+        if check_obj:
+            module = self.import_module(module_name)
+            obj2, _ = _getattribute(module, name)
+            self.check_obj(obj, obj2, name)
+
+        return module_name, name
+
+    def is_default(self) -> bool:
+        """Returns true if this ModuleEnv only represents the default Python environment."""
+        return self._importers == [DefaultImporter]
+
+    def check_obj(self, obj, obj2, name=None):
+        """Checks that two objects are the same identity.
+
+        Don't use this method, prefer to use the automatic check in `get_name`.
+        It's only separate to help implement CustomImportPickler.
+
+        Args:
+            obj: The object that the user gave us. This will be treated as such in error messaging.
+            obj2: The object retrieve by looking up a name in `self`.
+            name: If set, use this name instead of looking up __name__ or __qualname__ in error reporting.
+                This is only here to match how Pickler handles __reduce__ functions that return a string,
+                don't use otherwise.
+
+        Returns:
+            None
+
+        Raises:
+            ModuleEnvError: The two objects are not the same identity. The exception message will contain
+                a description and suggestions for fixing the issue.
+        """
+        if obj is obj2:
+            return
+
+        if name is None:
+            name = getattr(obj, "__qualname__", None)
+        if name is None:
+            name = obj.__name__
+
+        orig_module_name = self._whichmodule(obj, name)
+        obj_module_name = getattr(obj, "__module__", orig_module_name)
```

Comment:
from here to the end of the function is an intertwined copy/paste. Can it be refactored into a function that returns `mangled, obj_location, obj_importer_name`